primitives/arithmetic/src/traits.rs

```diff
@@ -302,3 +307,519 @@ pub trait SaturatedConversion {
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_add(self, v: Self) -> Result<Self, ArithmeticError> {
+			self.checked_add(&v).ok_or_else(|| error::equivalent(v))
+		}
+	}
+
+	/// Performs subtraction that returns `ArithmeticError` instead of wrapping around on underflow.
+	pub trait EnsureSub: CheckedSub + PartialOrd + Zero + Copy {
+		/// Subtracts two numbers, checking for overflow.
+		/// If overflow happens, `ArithmeticError` is returned.
+		///
+		/// Similar to [`CheckedSub::checked_sub()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureSub, ArithmeticError};
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     0u32.ensure_sub(1)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     i32::MAX.ensure_sub(-1)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// ```
+		fn ensure_sub(self, v: Self) -> Result<Self, ArithmeticError> {
+			self.checked_sub(&v).ok_or_else(|| error::inverse(v))
+		}
+	}
+
+	/// Performs multiplication that returns `ArithmeticError` instead of wrapping around on
+	/// overflow.
+	pub trait EnsureMul: CheckedMul + PartialOrd + Zero + Copy {
+		/// Multiplies two numbers, checking for overflow. If overflow happens,
+		/// `ArithmeticError` is returned.
+		///
+		/// Similar to [`CheckedMul::checked_mul()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureMul, ArithmeticError};
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     u32::MAX.ensure_mul(2)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     i32::MAX.ensure_mul(-2)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_mul(self, v: Self) -> Result<Self, ArithmeticError> {
+			self.checked_mul(&v).ok_or_else(|| error::multiplication(self, v))
+		}
+	}
+
+	/// Performs division that returns `ArithmeticError` instead of wrapping around on overflow.
+	pub trait EnsureDiv: CheckedDiv + PartialOrd + Zero + Copy {
+		/// Divides two numbers, checking for overflow.
+		/// If overflow happens, `ArithmeticError` is returned.
+		///
+		/// Similar to [`CheckedDiv::checked_div()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureDiv, ArithmeticError, FixedI64};
+		///
+		/// fn extrinsic_zero() -> Result<(), ArithmeticError> {
+		///     1.ensure_div(0)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     FixedI64::from(i64::MIN).ensure_div(FixedI64::from(-1))?;
```

Comment:
The range of i64 is (-9223372036854775808, 9223372036854775807) because is a two's-complement. This means that if we transform the min to positive, it will be out of range.

Reply:
Oh yes, thanks!